src/encrypt.rs

```diff
@@ -0,0 +1,302 @@
+        seq.serialize_element(&self.protected)?;
+        seq.serialize_element(&self.unprotected)?;
+        seq.serialize_element(&self.ciphertext)?;
+        seq.end()
+    }
+}
+
+///  Implementation of the Enc_structure structure as defined in
+///  [RFC8152](https://tools.ietf.org/html/rfc8152#section-5.3).
+///
+#[derive(Debug, Clone, Deserialize)]
+struct EncStructure {
+    /// context: "Encrypt0" / "Encrypt" / "Enc_Recipient" / "Mac_Recipient" / "Rec_Recipient"
+    context: String,
+    /// protected : empty_or_serialized_map,
+    protected: ByteBuf,
+    /// external_aad : bstr,
+    external_aad: ByteBuf,
+}
+
+impl Serialize for EncStructure {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(3))?;
+        seq.serialize_element(&self.context)?;
+        seq.serialize_element(&self.protected)?;
+        seq.serialize_element(&self.external_aad)?;
+        seq.end()
+    }
+}
+
+impl EncStructure {
+    ///
+    pub fn new_encrypt0(protected: &[u8]) -> Result<Self, CborError> {
+        Ok(EncStructure{
+            context: String::from("Encrypt0"),
+            protected: ByteBuf::from(protected.to_vec()),
+            external_aad: ByteBuf::new(),
+        })
+    }
+
+    /// Serializes the EncStructure to . We don't care about deserialization, since
+    /// both sides are supposed to compute the EncStructure and compare.
+    pub fn as_bytes(&self) -> Result<Vec<u8>, CborError> {
+        serde_cbor::to_vec(self)
+    }
+}
+
+impl COSEEncrypt0 {
+    ///
+    pub fn new(
+        payload: &[u8],
+        cipher: Cipher,
+        key: &PKeyRef<Private>,
+    ) -> Result<Self, COSEError> {
+        let mut iv = vec![0; cipher.iv_len().unwrap()];
+        rand_bytes(&mut iv).unwrap();
+
+        // https://datatracker.ietf.org/doc/html/rfc8152#section-3.1
+        // https://datatracker.ietf.org/doc/html/rfc8152#section-10.1
+        let mut protected = HeaderMap::new();
+        let cose_alg = match COSEAlgorithm::from_nid(cipher.nid()) {
+            None => return Err(COSEError::UnsupportedError(
+                "Unsupported encryption algorithm"
+                    .to_string(),
+            )),
+            Some(v) => v
+        };
+        let cose_alg_value = cose_alg.value();
+        protected.insert(1.into(), CborValue::Integer(cose_alg_value as i128));
+        let mut unprotected = HeaderMap::new();
+        unprotected.insert(5.into(), CborValue::Bytes(iv.to_owned()));
+
+        let protected_bytes =
+            map_to_empty_or_serialized(&protected).map_err(COSEError::SerializationError)?;
+
+        let enc_structure = EncStructure::new_encrypt0(&protected_bytes)
+            .map_err(COSEError::SerializationError)?;
+
+        let mut tag = vec![0; cose_alg.tag_size()];
+        let mut ciphertext = encrypt_aead(cipher,
+            &key.private_key_to_der().unwrap(),
```

Comment:
The `key` here should be the `&[u8]` directly.

Reply:
uhm isn't getting the der bytes enough? I've been digging into PKeyRef/PKey and couldn't find anything otherwise

Reply:
`PKey` is for `Public/Private keys`, i.e. asymmetric cryptography.

Reply:
changed the fns to accept an `&[u8]`